Cython/Compiler/ExprNodes.py

```diff
@@ -9371,21 +9371,22 @@ def analyse_default_args(self, env):
         # so their optional arguments must be static, too.
         # TODO: change CyFunction implementation to pass both function object and owning object for method calls
         must_use_constants = env.is_c_class_scope or (self.def_node.is_wrapper and env.is_module_scope)
-
         for arg in self.def_node.args:
-            if arg.default and not must_use_constants:
-                if not arg.default.is_literal:
-                    arg.is_dynamic = True
-                    if arg.type.is_pyobject:
-                        nonliteral_objects.append(arg)
+            if arg.default:
+                if not must_use_constants:
+                    if not arg.default.is_literal:
+                        arg.is_dynamic = True
+                        if arg.type.is_pyobject:
+                            nonliteral_objects.append(arg)
+                        else:
+                            nonliteral_other.append(arg)
                     else:
-                        nonliteral_other.append(arg)
-                else:
-                    arg.default = DefaultLiteralArgNode(arg.pos, arg.default)
-                if arg.kw_only:
-                    default_kwargs.append(arg)
-                else:
-                    default_args.append(arg)
+                        arg.default = DefaultLiteralArgNode(arg.pos, arg.default)
+                if arg.type.is_pyobject or arg.type.is_numeric:
```

Comment:
Why only these two? Why is this constraint needed?

Reply:
This was added to make sure that only primitives/pyobjects (which also includes `unicode` and `str`) could be added to `__(kw)defaults__` which would  block cases like `cpdef f(char* arg = NULL)`. 

Alternative would be to do nothing and in which case this example will end up being compile time error but it might break some existing code  that used to compile  fine before.

Reply:
I think `arg.type` should have a function `can_coerce_to_pyobject`?

Reply:
I think I've tried to use it however it does not  return what I would expect for `CPtrType` - in `CType` `can_coerce_to_pyobject` [is defined as](https://github.com/cython/cython/blob/master/Cython/Compiler/PyrexTypes.py#L1635) 
```
    def create_from_py_utility_code(self, env):
        return self.from_py_function is not None

    def can_coerce_to_pyobject(self, env):
        return self.create_to_py_utility_code(env)
```
and for `CPtrType` `from_py_function` will not be `None`. As a result `can_coerce_to_pyobject` for `CPtrType` returns `True` which lead to compile time error downstream.

Reply:
Does `arg.default` have a type? I think it may sometimes be different from `arg.type` (e.g. null pointers)

Reply:
Yes, in this case it is `<CPtrType <CVoidType>>` (code in question is `cpdef f(char* arg = NULL)`)